modules/geo-layers/src/imagery-layer/imagery-layer.ts

```diff
@@ -0,0 +1,208 @@
+      }
+
+      // Some sublayer props may have changed
+    }
+
+    if (changeFlags.viewportChanged) {
+      this.debounce(() => this.loadImage('viewport changed'));
+    }
+  }
+
+  override finalizeState(): void {
+    // TODO - we could cancel outstanding requests
+  }
+
+  override renderLayers(): Layer {
+    // TODO - which bitmap layer is rendered should depend on the current viewport
+    // Currently Studio only uses one viewport
+    const {bounds, image} = this.state;
+
+    return (
+      image &&
+      new BitmapLayer({
+        ...this.getSubLayerProps({id: 'bitmap'}),
+        bounds,
+        image
+      })
+    );
+  }
+
+  async getFeatureInfoText(x: number, y: number): Promise<string | null> {
+    const {viewport} = this.context;
+    if (viewport) {
+      const bounds = viewport.getBounds();
+      const {width, height} = viewport;
+      // @ts-expect-error Undocumented method
+      const featureInfo = await this.state.imageSource.getFeatureInfoText?.({
+        layers: this.props.layers,
+        // todo image width may get out of sync with viewport width
+        width,
+        height,
+        bbox: bounds,
+        query_layers: this.props.layers,
+        x,
+        y,
+        info_format: 'application/vnd.ogc.gml'
+      });
+      return featureInfo;
+    }
+    return '';
+  }
+
+  _createImageSource(props: ImageryLayerProps): ImageSource {
+    if (props.data instanceof ImageSource) {
+      return props.data;
+    }
+
+    if (typeof props.data === 'string') {
+      return createImageSource({
+        url: props.data,
+        loadOptions: props.loadOptions,
+        type: props.serviceType
+      });
+    }
+
+    throw new Error('invalid image source in props.data');
+  }
+
+  /** Run a getMetadata on the image service */
+  async _loadMetadata(): Promise<void> {
+    this.props.onMetadataLoadStart();
+    try {
+      this.state.metadata = await this.state.imageSource.getMetadata();
+      // technically we should get the latest layer after an async operation in case props have changed
+      // Although the response might no longer be expected
+      this.getCurrentLayer()?.props.onMetadataLoadComplete(this.state.metadata);
+    } catch (error) {
+      this.getCurrentLayer()?.props.onMetadataLoadError(error as Error);
+    }
+  }
+
+  /** Load an image */
+  async loadImage(reason: string): Promise<void> {
+    // TODO - need to handle multiple viewports
+    const {viewport} = this.context;
+    if (!viewport) {
+      return;
+    }
+    const bounds = viewport.getBounds();
+    const {width, height} = viewport;
+
+    const requestId = this.getRequestId();
+
+    try {
+      this.props.onImageLoadStart(requestId);
+      const image = await this.state.imageSource.getImage({
+        width,
+        height,
+        bbox: bounds,
+        layers: this.props.layers
+      });
+      Object.assign(this.state, {image, bounds, width, height});
```

Comment:
Call `this.setState`. The method will also trigger `setNeedsRedraw`

Reply:
I think you'll need to check that `requestId` is larger than the last loaded image before overwriting the state.